tokio/src/io/driver/mod.rs

```diff
@@ -47,6 +49,14 @@ pub(crate) struct ReadyEvent {
 }
 
 pub(super) struct Inner {
+    /// Primary slab handle containing the state for each resource registered
+    /// with this driver.
+    ///
+    /// The ownership of this slab is moved into this structure during
+    /// `Driver::drop`, so that `Inner::drop` can notify all outstanding handles
+    /// without risking new ones being registered in the meantime.
+    resources: Mutex<Option<Slab<ScheduledIo>>>,
```

Comment:
Moving it into `Inner` isn't what ensures no further resources are registered. Because `Handle` holds a weak ref to `Inner`, we know that, when `Inner` drops, no further registration can happen. So cleanup happens at that point in `Inner::drop`.

Reply:
I do think there probably is a better way to handle shutdown, but this is the quickest way to correct the issues that exist today w/o a bigger refactor. Impact is minimal as well as the `Mutex` is only used once in the lifecycle of the I/O driver (shutdown).

Reply:
God it, that makes sense! Thank you. Out of curiosity, what would that better way look like?